lazy-pages/src/sys.rs

```diff
@@ -39,117 +53,284 @@ cfg_if! {
+
+    // Is definitely write if it's second access or there `is_write` is set as `true`.
+    let is_psg_case = is_definitely_write
+        && native_ps < psg
+        && !sp_io::storage::exists(prefix.calc_key_for_page(wasm_addr.as_page_number()));
+    let unprot_size = if is_psg_case {
+        log::trace!("is PSG case - we need to upload to storage data for all pages from `page-storage-granularity`");
+        wasm_addr.align_down(psg);
+        psg
     } else {
-        assert_eq!(gear_ps % native_ps, 0);
-        (gear_page, 1usize, wasm_mem_begin + gear_page.offset())
+        native_ps
     };
 
-    let accessed_page = PageNumber(((mem as usize - wasm_mem_begin) / gear_ps) as u32);
-    log::debug!(
-        "mem={:?} accessed={:?},{:?} pages={:?} page_native_addr={:#x}",
-        mem,
-        accessed_page,
-        accessed_page.to_wasm_page(),
-        gear_page.0..gear_page.0 + gear_pages_num as u32,
-        unprot_addr
-    );
+    wasm_addr.check_interval_with_size_is_ok(unprot_size)?;
+
+    // Set r/w protection in order to load data from storage into mem buffer.
+    let unprot_addr = wasm_addr.as_native_addr();
+    log::trace!("mprotect r/w, addr = {unprot_addr:#x}, size = {unprot_size:#x}");
+    region::protect(
+        unprot_addr as *mut (),
+        unprot_size as usize,
+        Protection::READ_WRITE,
+    )?;
 
-    let unprot_size = gear_pages_num * gear_ps;
-
-    region::protect(unprot_addr as *mut (), unprot_size, Protection::READ_WRITE)?;
-
-    LAZY_PAGES_CONTEXT.with(|ctx| {
-        let mut ctx = ctx.borrow_mut();
-        for idx in 0..gear_pages_num as u32 {
-            let page = gear_page + idx.into();
-
-            let ptr = (unprot_addr as *mut u8).add(idx as usize * gear_ps);
-            let buffer_as_slice = std::slice::from_raw_parts_mut(ptr, gear_ps);
-
-            // TODO: simplify before release (issue #1147). Currently we must support here all old runtimes.
-            // For new runtimes we have to calc page key from program pages prefix.
-            let page_key = if let Some(prefix) = &ctx.program_storage_prefix {
-                page_key_in_storage(prefix, page)
-            } else {
-                // This case is for old runtimes support
-                ctx.lazy_pages_info.remove(&page).ok_or(Error::LazyPageNotExistForSignalAddr(mem, page))?
-            };
-            let res = sp_io::storage::read(&page_key, buffer_as_slice, 0);
-
-            if res.is_none() {
-                log::trace!(
-                    "{:?} has no data in storage, so just save current page data to released pages",
-                    page
-                );
-            } else {
-                log::trace!("{:?} has data in storage, so set this data for page and save it in released pages", page);
+    let fist_gear_page = wasm_addr.as_page_number();
+
+    for idx in 0..unprot_size / lazy_page_size {
+        // Arith opertaions are safe here.
```

Comment:
`...because`?

Reply:
It's very dirty to explain each arith operation in code

Reply:
I mean can you write something to explain in general?

Reply:
Ok, make some new comments